src/core/search/ast_expr.h

```diff
@@ -4,21 +4,106 @@
 
 #pragma once
 
+#include <algorithm>
+#include <iostream>
+#include <memory>
 #include <ostream>
+#include <vector>
+
+#include "base/logging.h"
 
 namespace dfly {
 
 namespace search {
 
-class AstExpr {};
+// Describes a single node of the filter AST tree.
+class AstNode {
+ public:
+  virtual ~AstNode() = default;
+
+  // Check if this input is matched by the node.
+  virtual bool Check(std::string_view input) const = 0;
+
+  // Debug print node.
+  virtual std::string Debug() const = 0;
+};
+
+using AstNodePtr = std::unique_ptr<AstNode>;
+
+// Do we have to pollute all our parser file with moves?
+struct AstExpr {
+  AstExpr() : ptr{} {
+  }
+  AstExpr(AstNodePtr ptr) : ptr{move(ptr)} {
```

Comment:
This is an unusual way to create a move constructor.
Standard move constructors accept an rvalue ref (`AstNodePtr&&`) and move from it, whereas copy c'tors (like the one below) have a `const AstExpr&` signature.
I'd do that, as it makes sure that calls with temporary (rvalue) objects are using the correct c'tor. I'm not sure if overload resolution would call this c'tor or the one below, as it's a subtle definition.